shared/ssz/decode.go

```diff
@@ -0,0 +1,220 @@
+package ssz
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+)
+
+// TODO(1068): Support more data types
+
+// Decode decodes data read from r and output it into the object pointed by pointer val
+func Decode(r io.Reader, val interface{}) error {
+	return decode(r, val)
+}
+
+func decode(r io.Reader, val interface{}) error {
+	if val == nil {
+		return newDecodeError("cannot decode into nil", nil)
+	}
+	rval := reflect.ValueOf(val)
+	rtyp := rval.Type()
+	if rtyp.Kind() != reflect.Ptr {
+		return newDecodeError("can only decode into pointer target", rtyp)
+	}
+	if rval.IsNil() {
+		return newDecodeError("cannot output to pointer of nil", rtyp)
+	}
+	encDec, err := cachedEncoderDecoder(rval.Elem().Type())
+	if err != nil {
+		return newDecodeError(fmt.Sprint(err), rval.Elem().Type())
+	}
+	if _, err = encDec.decoder(r, rval.Elem()); err != nil {
+		return newDecodeError(fmt.Sprint(err), rval.Elem().Type())
+	}
+	return nil
+}
+
+func makeDecoder(typ reflect.Type) (dec decoder, err error) {
+	kind := typ.Kind()
+	switch {
+	case kind == reflect.Bool:
+		return decodeBool, nil
+	case kind == reflect.Uint8:
+		return decodeUint8, nil
+	case kind == reflect.Uint16:
+		return decodeUint16, nil
+	case kind == reflect.Slice && typ.Elem().Kind() == reflect.Uint8:
+		return decodeBytes, nil
+	case kind == reflect.Slice:
+		return makeSliceDecoder(typ)
+	case kind == reflect.Struct:
+		return makeStructDecoder(typ)
+	default:
+		return nil, fmt.Errorf("type %v is not deserializable", typ)
+	}
+}
+
+func decodeBool(r io.Reader, val reflect.Value) (uint32, error) {
+	b := make([]byte, 1)
+	if err := readBytes(r, 1, b); err != nil {
+		return 0, err
+	}
+	v := uint8(b[0])
+	if v == 0 {
+		val.SetBool(false)
+	} else if v == 1 {
+		val.SetBool(true)
+	} else {
+		return 0, errors.New("expect 0 or 1 for decoding bool")
```

Comment:
add in the value received instead of 0 and 1 :
```go
return 0 , fmt.Errorf("expected 0 or 1 for decoding bool but got %d",v)
```

Reply:
Great suggestion! Fixed